src/voter.rs

```diff
@@ -803,63 +831,132 @@ impl<H, N, E: Environment<H, N>> Voter<H, N, E> where
+			let prospective_round_number =
+				self.prospective_round.as_ref().map(|r| r.votes.number());
+
+			// we saw a commit for a round `r` that is higher than our current best
+			// round so we start a prospective round at `r + 1`
+			if round_number > self.best_round.votes.number() + 1 &&
+				prospective_round_number.iter().all(|n| round_number > *n) {
+					trace!(target: "afg", "Imported commit for later round than current best {}, starting prospective round at {}",
+						   self.best_round.votes.number(),
+						   round_number + 1);
+
+					// we set `last_round_state` to `None` so that no votes are cast
+					self.prospective_round = Some(VotingRound::new(
+						round_number + 1,
+						(commit.target_hash.clone(), commit.target_number.clone()),
+						None,
+						self.best_round.finalized_sender.clone(),
+						self.env.clone(),
+					));
+				}
 		}
-		false
+
+		Ok(())
 	}
-}
 
-impl<H, N, E: Environment<H, N>> Future for Voter<H, N, E> where
-	H: Hash + Clone + Eq + Ord + ::std::fmt::Debug,
-	N: Copy + BlockNumberOps + ::std::fmt::Debug,
-{
-	type Item = ();
-	type Error = E::Error;
+	// Processes the current prospective round, returns `Async::NotReady` when
+	// the `best_round` has been updated and the caller should re-poll.
+	fn process_prospective_round(&mut self) -> Poll<Option<bool>, E::Error> {
+		let mut best_round_completable = None;
 
-	fn poll(&mut self) -> Poll<(), E::Error> {
-		self.prune_background()?;
-		self.committer.poll()?;
+		// poll the prospective round (if any),
+		if let Some(mut prospective_round) = self.prospective_round.take() {
+			// if it is completable then we background it and start the new
+			// `best_round` at `prospective_round.number + 1`.
+			if let Async::Ready(()) = prospective_round.poll()? {
+				trace!(target: "afg", "Prospective round at {} has become completable. Starting best round at {}",
+					   prospective_round.votes.number(),
+					   prospective_round.votes.number() + 1);
+
+				self.completed_prospective_round(prospective_round)?;
+
+				// round has been updated, so we return `NotReady` to trigger a re-poll.
+				return Ok(Async::NotReady);
+
+			} else {
+				assert!(self.best_round.votes.number() < prospective_round.votes.number());
+				if let Async::Ready(()) = self.best_round.poll()? {
+					// if `best_round` is completable and we've caught up with prospective round,
+					// then we background the current `best_round` and set the `best_round`
+					// to `prospective_round`.
+					if self.best_round.votes.number() == prospective_round.votes.number() - 1 {
+						trace!(target: "afg", "Best round at {} has caught up with prospective round at {}. \
+											   Setting best round to prospective round.",
+							   self.best_round.votes.number(),
+							   prospective_round.votes.number());
+
+						prospective_round.last_round_state = Some(self.best_round.bridge_state());
+						self.completed_best_round(Some(prospective_round))?;
```

Comment:
the `prospective_round` isn't completed here, is it? The `best_round` is though. Also it doesn't look like the get swapped anymore

Reply:
ahh, I misread; I see now that `completed_best_round` takes the next round.